pkg/pillar/utils/cloudconfig/cloudconfig.go

```diff
@@ -0,0 +1,88 @@
+package cloudconfig
+
+import (
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+
+	"github.com/lf-edge/eve/pkg/pillar/base"
+	fileutils "github.com/lf-edge/eve/pkg/pillar/utils/file"
+	"gopkg.in/yaml.v2"
+)
+
+type CloudConfig struct {
+	RunCmd     []string       `yaml:"runcmd"`
+	WriteFiles []WritableFile `yaml:"write_files"`
+}
+
+type WritableFile struct {
+	Path        string `yaml:"path"`
+	Content     string `yaml:"content"`
+	Permissions string `yaml:"permissions"`
+	Encoding    string `yaml:"encoding"`
+	Owner       string `yaml:"owner"`
+}
+
+func IsCloudConfig(ci string) bool {
+	// check if the first line is #cloud-config
+	lines := strings.Split(ci, "\n")
+	if len(lines) == 0 {
+		return false
+	}
+	return strings.HasPrefix(lines[0], "#cloud-config")
+}
+
+func ParseCloudConfig(ci string) (*CloudConfig, error) {
+	var cc CloudConfig
+	err := yaml.Unmarshal([]byte(ci), &cc)
+	if err != nil {
+		return nil, err
+	}
+	return &cc, nil
+}
+
+func WriteFile(log *base.LogObject, file WritableFile, rootPath string) error {
+	// transform file.Permission to os.FileMode
+	perm, err := strconv.ParseUint(file.Permissions, 8, 32)
+	if err != nil {
+		return err
+	}
+	mode := os.FileMode(perm)
+
+	writePath := filepath.Join(rootPath, file.Path)
+	// sanitize path
+	if !strings.HasPrefix(filepath.Clean(writePath), rootPath) {
+		return fmt.Errorf("invalid path %s", writePath)
```

Comment:
I think it would be better to log file.Path and maybe have the message be that the path tries to escape the root.